[Validator] Allow objects implementing __toString() to be used as violation messages

mdlutz24 opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | I didn't see a doc on violations to update, but I'm happy to do documentation if somone can suggest the best place to do it.

Currently in the Drupal project we use Translatable Markup object to hold most strings and currently pass them in as Constraint Violation messages. In Symfony 3 this works but with the added typehinting in Symfony 4, these markup objects are rendered into strings at the time of the violation creation.  This causes any html in the message string to be considered unsafe by twig later in our rendering process.  This pr explicitly allows objects implementing a __toString() method to be used as violation messages, and the violation will save and return the original stringable object.

See https://www.drupal.org/project/drupal/issues/3029540 For our Drupal issue on the subject.

ro0NL commented on src/Symfony/Component/Validator/ConstraintViolationInterface.php:
```diff
@@ -36,7 +36,7 @@ interface ConstraintViolationInterface
      * Returns the violation message.
      *
-     * @return string The violation message
+     * @return mixed The violation message
```
i guess this shows you relied on implicit behavior :) we cant change it because of BC. So perhaps cast on read in its implementation?

But in general a message is a string, i tend to believe it should be passed as is.

mdlutz24 replied:
ro0NL - Thank you for reviewing this. I appreciate your input.

We've considered recasting on read. We can convert it back to a Markup object to make twig consider the html safe, but we can't store enough information in the violation to recreate the original Translatable Markup object that we use, so the translation happens upon constructing the violation rather than in the twig layer, which is not ideal.

Expanding the allowables types for the constructor is not a BC break, but expanding the return type of getMessage from string to mixed would be, but I would argue that the break is mitigated by the fact that 1) the return type is not technically a part of the interface (yet) 2) the return value in this PR is guaranteed to be castable to a string and 3) you would only get an object if you passed one in in the first place.

stof replied:
> 1) the return type is not technically a part of the interface (yet)

that's wrong. Our BC policy covers the return type, even when only phpdoc is used to define it.

nicolas-grekas replied:
let's keep it as `string` but alter the description instead, eg
`The violation message as a string or a stringable object`

alexpott commented:
@michaelcullum and I talked a bit around this issue and whether or not we should be pushing for a Stringable pseudo-class in PHP - there are prior discussions - see https://externals.io/message/98424

xabbuh commented:
I am :-1: on making this change as it calls for breaking existing use cases that do not perform explicit string casts on the returned value of `getMessage()`.

nicolas-grekas commented:
What are the alternatives on your side @alexpott @mdlutz24?

alexpott commented:
@nicolas-grekas none of them are particularly pretty - https://www.drupal.org/project/drupal/issues/3029540 - we've explored quite a few there.

ro0NL commented:
does the issue affect drupal users directly? (if not i think drupal core should just patch)

e.g. https://github.com/symfony/symfony/blob/59c08eda702b4f07af9b30377ebb10cb55123797/src/Symfony/Component/Validator/Context/ExecutionContextInterface.php#L67-L70

is the issue drupal users do `add/buildViolation(t('message'))` at this point? Is it possible for drupal to provide its own `ExecutionContextInterface`?

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintViolation.php:
```diff
@@ -32,7 +34,7 @@ class ConstraintViolation implements ConstraintViolationInterface
      * Creates a new constraint violation.
      *
-     * @param string          $message         The violation message
+     * @param mixed           $message         The violation message
```
let's keep it as `string` but alter the description instead, eg
`The violation message as a string or a stringable object`

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintViolation.php:
```diff
@@ -49,8 +51,11 @@ class ConstraintViolation implements ConstraintViolationInterface
+    public function __construct($message, ?string $messageTemplate, array $parameters, $root, ?string $propertyPath, $invalidValue, int $plural = null, $code = null, Constraint $constraint = null, $cause = null)
     {
+        if (!(null === $message || \is_string($message) || (\is_object($message) && method_exists($message, '__toString')))) {
+            throw new InvalidArgumentException('Constraint violation message should be a string or an object which implements the __toString() method.');
```
let's use `\TypeError`

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintViolation.php:
```diff
@@ -49,8 +51,11 @@ class ConstraintViolation implements ConstraintViolationInterface
-    public function __construct(?string $message, ?string $messageTemplate, array $parameters, $root, ?string $propertyPath, $invalidValue, int $plural = null, $code = null, Constraint $constraint = null, $cause = null)
+    public function __construct($message, ?string $messageTemplate, array $parameters, $root, ?string $propertyPath, $invalidValue, int $plural = null, $code = null, Constraint $constraint = null, $cause = null)
     {
+        if (!(null === $message || \is_string($message) || (\is_object($message) && method_exists($message, '__toString')))) {
```
`if (!\is_string($message ?? '') && !(\is_object($message) && method_exists($message, '__toString'))) {`

mdlutz24 replied:
It appears that passing a NULL message has been deprecated since I originally sent this PR, and the handling code for that makes both the null === $message and the null coalesce here unnecessary in the current code, as $message cannot be null.

However, the handling code that deprecated passing null as a message  was under the assumption that there would be a 'string' typehint beginning in symfony 5, and if we are going to allow stringable objects, then we cannot typehint a string there.

nicolas-grekas reviewed (commented):
We have to do this change. Let's be friendly to our users. It's the only pragmatic move. 
Here are some CS changes before approving on my side.

nicolas-grekas commented:
(please rebase also, on branch 4.4)

ro0NL commented:
note in 5.0 there are more typehints that could be blocking still;

https://github.com/symfony/symfony/blob/1f8f4f2011c0fa93515d3141cbbf77a1b88aeb57/src/Symfony/Component/Validator/Context/ExecutionContextInterface.php#L70

https://github.com/symfony/symfony/blob/1f8f4f2011c0fa93515d3141cbbf77a1b88aeb57/src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php#L46

nicolas-grekas commented:
@ro0NL thanks, then let's add `as a string or a stringable object` to all places where this would block lazy messages. IMHO.

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintViolation.php:
```diff
@@ -60,6 +60,10 @@ public function __construct(?string $message, ?string $messageTemplate, array $p
             @trigger_error(sprintf('Not using a string as the error code in %s() is deprecated since Symfony 4.4. A type-hint will be added in 5.0.', __METHOD__), E_USER_DEPRECATED);
         }
 
+        if (!(\is_string($message) || (\is_object($message) && method_exists($message, '__toString')))) {
```
`if (!\is_string($message) && !(\is_object($message) && method_exists($message, '__toString'))) {`

nicolas-grekas commented on src/Symfony/Component/Validator/ConstraintViolationInterface.php:
```diff
@@ -36,7 +36,7 @@ interface ConstraintViolationInterface
      * Returns the violation message.
      *
-     * @return string The violation message
+     * @return string The violation message as a string or a stringable object
```
let's do the same on https://github.com/symfony/symfony/pull/31083#issuecomment-516569294 if you can confirm it's required with your issue to be solved?

mdlutz24 replied:
Done.

nicolas-grekas reviewed (commented):
please have a look at tests :)

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -36,6 +36,7 @@ CHANGELOG
  * added `PositiveOrZero` constraint
  * added `Negative` constraint
  * added `NegativeOrZero` constraint
+ * added ability to use stringable objects as violation messages
```
should be moved to 4.4.0 section

mdlutz24 replied:
Done.

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -47,7 +47,7 @@ class ConstraintViolationBuilder implements ConstraintViolationBuilderInterface
     /**
      * @param TranslatorInterface $translator
```
we should add `@param string $message The error message as a string or a stringable object` here too, to not forget about it

mdlutz24 replied:
And Done.

nicolas-grekas commented on src/Symfony/Component/Validator/Violation/ConstraintViolationBuilder.php:
```diff
@@ -46,8 +46,9 @@ class ConstraintViolationBuilder implements ConstraintViolationBuilderInterface
 
     /**
      * @param TranslatorInterface $translator
+     * @param string              $message    The error message as a string or a stringable object
```
please put it before $translator, to respect the declaration order

mdlutz24 replied:
Done.  Thanks for all the reviews.  This would be my first contribution to Symfony, so I appreciate the guidance.

nicolas-grekas reviewed (approved):
Good on my side (just one very minor CS comment)
Thanks for being so reactive.

nicolas-grekas commented:
(rebase needed)

mdlutz24 committed:
[Validator] Allow objects implementing __toString() to be used as violation messages

[Validator] updated changelog

[Validator] updated typehint for ConstraintViolationInterface::getMessage()

[Validator] fixed spacing issue inadvertantly added in previous commit

[Validator] fixed coding standard issues

[Validator] Address feedback

[Validator] Fix coding standard violation

[Validator] update tests

[Validator] Address feedback, Round 2

[Validator] Document ConstraintViolationBuilder::__construct()  parameter

[Validator] Update changelog

[Validator] Adjust parameter documentation order

mdlutz24 commented:
Rebased and squashed

fabpot commented:
Thank you @mdlutz24.
